File: sdks/python/quota/extractors.py
from typing import Dict, Any, Optional, Tuple, List

def detect_provider(url: str) -> Optional[str]:
    """Identifica o provedor de IA com base na URL de destino da requisição."""
    lowercase_url = url.lower()
    if "api.openai.com" in lowercase_url:
        return "openai"
    elif "api.anthropic.com" in lowercase_url:
        return "anthropic"
    elif "generativelanguage.googleapis.com" in lowercase_url:
        return "google"
    elif "api.groq.com" in lowercase_url:
        return "groq"
    elif "api.mistral.ai" in lowercase_url:
        return "mistral"
    elif "api.together.xyz" in lowercase_url:
        return "together"
    elif "api.cohere.com" in lowercase_url:
        return "cohere"
    return None

def extract_metadata_from_headers(headers: Any) -> Dict[str, Any]:
    """Extrai metadados customizados a partir dos cabeçalhos HTTP da requisição (ex: x-quota-agent)."""
    if not headers:
        return {}

    # Normaliza headers para um dicionário case-insensitive
    normalized_headers: Dict[str, str] = {}
    if hasattr(headers, "items"):
        for k, v in headers.items():
            normalized_headers[str(k).lower()] = str(v)
    elif isinstance(headers, (list, tuple)):
        for item in headers:
            if isinstance(item, (list, tuple)) and len(item) == 2:
                normalized_headers[str(item[0]).lower()] = str(item[1])

    get_val = lambda name: normalized_headers.get(name.lower())

    project = get_val("x-quota-project")
    agent = get_val("x-quota-agent")
    environment = get_val("x-quota-environment")
    external_user_id = get_val("x-quota-user-id")
    request_group = get_val("x-quota-request-group")
    billing_group = get_val("x-quota-billing-group")
    tags_str = get_val("x-quota-tags")

    tags = [t.strip() for t in tags_str.split(",") if t.strip()] if tags_str else None

    result: Dict[str, Any] = {}
    if project: result["project"] = project
    if agent: result["agent"] = agent
    if environment: result["environment"] = environment
    if external_user_id: result["externalUserId"] = external_user_id
    if request_group: result["requestGroup"] = request_group
    if billing_group: result["billingGroup"] = billing_group
    if tags: result["tags"] = tags

    return result

def extract_telemetry(
    provider: str,
    request_body: Optional[Dict[str, Any]],
    response_body: Optional[Dict[str, Any]],
    latency_ms: int,
    status_code: int,
    default_metadata: Optional[Dict[str, Any]] = None,
    header_metadata: Optional[Dict[str, Any]] = None
) -> Dict[str, Any]:
    """Extrai os dados de telemetria e contagem de tokens."""
    request_body = request_body or {}
    response_body = response_body or {}
    default_metadata = default_metadata or {}
    header_metadata = header_metadata or {}

    model = response_body.get("model") or request_body.get("model") or "unknown"

    prompt_tokens = 0
    completion_tokens = 0
    total_tokens = 0
    cached_tokens = 0
    reasoning_tokens = 0
    cache_creation_tokens = 0

    if provider in ["openai", "groq", "together", "mistral"]:
        usage = response_body.get("usage", {})
        if usage:
            prompt_tokens = usage.get("prompt_tokens", 0)
            completion_tokens = usage.get("completion_tokens", 0)
            total_tokens = usage.get("total_tokens", prompt_tokens + completion_tokens)
            prompt_details = usage.get("prompt_tokens_details", {}) or {}
            cached_tokens = prompt_details.get("cached_tokens", 0)
            comp_details = usage.get("completion_tokens_details", {}) or {}
            reasoning_tokens = comp_details.get("reasoning_tokens", 0)
    elif provider == "anthropic":
        usage = response_body.get("usage", {})
        if usage:
            prompt_tokens = usage.get("input_tokens", 0)
            completion_tokens = usage.get("output_tokens", 0)
            total_tokens = prompt_tokens + completion_tokens
            cached_tokens = usage.get("cache_read_input_tokens", 0)
            cache_creation_tokens = usage.get("cache_creation_input_tokens", 0)
    elif provider == "google":
        usage = response_body.get("usageMetadata", {})
        if usage:
            prompt_tokens = usage.get("promptTokenCount", 0)
            completion_tokens = usage.get("candidatesTokenCount", 0)
            total_tokens = usage.get("totalTokenCount", prompt_tokens + completion_tokens)
            cached_tokens = usage.get("cachedContentTokenCount", 0)
    else:
        usage = response_body.get("usage") or response_body.get("usageMetadata") or {}
        if usage:
            prompt_tokens = usage.get("prompt_tokens") or usage.get("promptTokenCount") or usage.get("input_tokens") or 0
            completion_tokens = usage.get("completion_tokens") or usage.get("candidatesTokenCount") or usage.get("output_tokens") or 0
            total_tokens = usage.get("total_tokens") or usage.get("totalTokenCount") or (prompt_tokens + completion_tokens)

    body_metadata = request_body.get("metadata") or {}
    merged_metadata = {
        "project": header_metadata.get("project") or body_metadata.get("project") or request_body.get("project") or default_metadata.get("project"),
        "agent": header_metadata.get("agent") or body_metadata.get("agent") or request_body.get("agent") or default_metadata.get("agent"),
        "environment": header_metadata.get("environment") or body_metadata.get("environment") or default_metadata.get("environment"),
        "externalUserId": header_metadata.get("externalUserId") or body_metadata.get("externalUserId"),
        "requestGroup": header_metadata.get("requestGroup") or body_metadata.get("requestGroup"),
        "tags": header_metadata.get("tags") or body_metadata.get("tags") or default_metadata.get("tags"),
    }

    # Remove chaves com valor None para limpar o JSON
    merged_metadata = {k: v for k, v in merged_metadata.items() if v is not None}

    return {
        "provider": provider,
        "model": model,
        "promptTokens": prompt_tokens,
        "completionTokens": completion_tokens,
        "totalTokens": total_tokens,
        "cachedTokens": cached_tokens,
        "reasoningTokens": reasoning_tokens,
        "cacheCreationTokens": cache_creation_tokens,
        "latencyMs": latency_ms,
        "statusCode": status_code,
        "success": 200 <= status_code < 300,
        "billingGroup": header_metadata.get("billingGroup") or request_body.get("billingGroup"),
        "metadata": merged_metadata
    }

File: sdks/js/src/extractors.ts
import { CollectorPayload, QuotaMetadata } from './types';

/**
 * Identifica o provedor de IA com base na URL de destino da requisição
 */
export function detectProvider(url: string): string | null {
  const lowercaseUrl = url.toLowerCase();
  if (lowercaseUrl.includes('api.openai.com')) return 'openai';
  if (lowercaseUrl.includes('api.anthropic.com')) return 'anthropic';
  if (lowercaseUrl.includes('generativelanguage.googleapis.com')) return 'google';
  if (lowercaseUrl.includes('api.groq.com')) return 'groq';
  if (lowercaseUrl.includes('api.mistral.ai')) return 'mistral';
  if (lowercaseUrl.includes('api.together.xyz')) return 'together';
  if (lowercaseUrl.includes('api.cohere.com')) return 'cohere';
  return null;
}

/**
 * Identifica o provedor a partir do nome do modelo (inferência automática)
 */
export function inferProviderFromModel(modelName?: string): string {
  if (!modelName) return 'openai';
  const m = modelName.toLowerCase();
  if (m.includes('gpt') || m.includes('o1') || m.includes('o3') || m.includes('dall-e') || m.includes('text-embedding')) return 'openai';
  if (m.includes('claude') || m.includes('anthropic')) return 'anthropic';
  if (m.includes('gemini') || m.includes('palm') || m.includes('gemma')) return 'google';
  if (m.includes('mistral') || m.includes('mixtral') || m.includes('codestral') || m.includes('pixtral')) return 'mistral';
  if (m.includes('groq') || m.includes('llama') || m.includes('gemma-7b')) return 'groq';
  return 'openai';
}

/**
 * Extrai metadados customizados a partir dos cabeçalhos HTTP da requisição (ex: x-quota-agent, x-quota-user-id)
 */
export function extractMetadataFromHeaders(headers?: HeadersInit): QuotaMetadata {
  if (!headers) return {};

  const getHeader = (name: string): string | undefined => {
    if (typeof Headers !== 'undefined' && headers instanceof Headers) {
      return headers.get(name) || undefined;
    } else if (Array.isArray(headers)) {
      const match = headers.find(([k]) => k.toLowerCase() === name.toLowerCase());
      return match ? match[1] : undefined;
    } else if (typeof headers === 'object') {
      const key = Object.keys(headers).find(k => k.toLowerCase() === name.toLowerCase());
      return key ? (headers as Record<string, string>)[key] : undefined;
    }
    return undefined;
  };

  const project = getHeader('x-quota-project');
  const agent = getHeader('x-quota-agent');
  const environment = getHeader('x-quota-environment');
  const externalUserId = getHeader('x-quota-user-id');
  const requestGroup = getHeader('x-quota-request-group');
  const billingGroup = getHeader('x-quota-billing-group');
  const tagsStr = getHeader('x-quota-tags');
  const tags = tagsStr ? tagsStr.split(',').map(t => t.trim()).filter(Boolean) : undefined;

  return {
    project,
    agent,
    environment,
    externalUserId,
    requestGroup,
    billingGroup,
    tags
  };
}

/**
 * Extrai os dados de telemetria a partir do payload da resposta e da requisição
 */
export function extractTelemetry(
  provider: string,
  requestBodyJson: any,
  responseBodyJson: any,
  latencyMs: number,
  statusCode: number,
  defaultMetadata?: QuotaMetadata,
  headerMetadata?: QuotaMetadata
): CollectorPayload {
  const model =
    responseBodyJson?.model ||
    requestBodyJson?.model ||
    'unknown';

  let promptTokens = 0;
  let completionTokens = 0;
  let totalTokens = 0;
  let cachedTokens = 0;
  let reasoningTokens = 0;
  let cacheCreationTokens = 0;

  // 1. OpenAI, Groq, Mistral, Together
  if (provider === 'openai' || provider === 'groq' || provider === 'together' || provider === 'mistral') {
    const usage = responseBodyJson?.usage;
    if (usage) {
      promptTokens = usage.prompt_tokens ?? 0;
      completionTokens = usage.completion_tokens ?? 0;
      totalTokens = usage.total_tokens ?? (promptTokens + completionTokens);
      cachedTokens = usage.prompt_tokens_details?.cached_tokens ?? 0;
      reasoningTokens = usage.completion_tokens_details?.reasoning_tokens ?? 0;
    }
  } 
  // 2. Anthropic (Claude)
  else if (provider === 'anthropic') {
    const usage = responseBodyJson?.usage;
    if (usage) {
      promptTokens = usage.input_tokens ?? 0;
      completionTokens = usage.output_tokens ?? 0;
      totalTokens = promptTokens + completionTokens;
      cachedTokens = usage.cache_read_input_tokens ?? 0;
      cacheCreationTokens = usage.cache_creation_input_tokens ?? 0;
    }
  } 
  // 3. Google Gemini
  else if (provider === 'google') {
    const usage = responseBodyJson?.usageMetadata;
    if (usage) {
      promptTokens = usage.promptTokenCount ?? 0;
      completionTokens = usage.candidatesTokenCount ?? 0;
      totalTokens = usage.totalTokenCount ?? (promptTokens + completionTokens);
      cachedTokens = usage.cachedContentTokenCount ?? 0;
      reasoningTokens = usage.thoughtsTokenCount ?? 0;
    }
  } 
  // Fallback genérico inteligente
  else {
    const usage = responseBodyJson?.usage || responseBodyJson?.usageMetadata;
    if (usage) {
      promptTokens = usage.prompt_tokens ?? usage.promptTokenCount ?? usage.input_tokens ?? usage.inputTokens ?? 0;
      completionTokens = usage.completion_tokens ?? usage.candidatesTokenCount ?? usage.output_tokens ?? usage.outputTokens ?? 0;
      totalTokens = usage.total_tokens ?? usage.totalTokenCount ?? (promptTokens + completionTokens);
      cachedTokens = usage.cached_tokens ?? usage.prompt_tokens_details?.cached_tokens ?? usage.cache_read_input_tokens ?? usage.cachedContentTokenCount ?? 0;
      reasoningTokens = usage.reasoning_tokens ?? usage.completion_tokens_details?.reasoning_tokens ?? usage.thoughtsTokenCount ?? 0;
    }
  }

  const bodyMetadata = requestBodyJson?.metadata || {};
  const mergedMetadata: QuotaMetadata = {
    project: headerMetadata?.project || bodyMetadata.project || requestBodyJson?.project || defaultMetadata?.project,
    agent: headerMetadata?.agent || bodyMetadata.agent || requestBodyJson?.agent || defaultMetadata?.agent,
    environment: headerMetadata?.environment || bodyMetadata.environment || defaultMetadata?.environment,
    externalUserId: headerMetadata?.externalUserId || bodyMetadata.externalUserId,
    requestGroup: headerMetadata?.requestGroup || bodyMetadata.requestGroup,
    billingGroup: headerMetadata?.billingGroup || requestBodyJson?.billingGroup,
    tags: headerMetadata?.tags || bodyMetadata.tags || defaultMetadata?.tags
  };

  return {
    provider,
    model,
    promptTokens,
    completionTokens,
    totalTokens,
    cachedTokens,
    reasoningTokens,
    cacheCreationTokens,
    latencyMs,
    statusCode,
    success: statusCode >= 200 && statusCode < 300,
    billingGroup: mergedMetadata.billingGroup,
    metadata: mergedMetadata
  };
}

/**
 * Extrai contagem de tokens de qualquer resposta de LLM (OpenAI, Anthropic Claude, Google Gemini, Groq, Mistral)
 * ou estruturas de ferramentas MCP (tools/call, sampling, createMessage)
 */
export function extractMcpTokens(responseData: any): {
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
  cachedTokens: number;
  reasoningTokens: number;
  model?: string;
} {
  if (!responseData) {
    return { promptTokens: 0, completionTokens: 0, totalTokens: 0, cachedTokens: 0, reasoningTokens: 0 };
  }

  const usage =
    responseData?.usage ||
    responseData?._meta?.usage ||
    responseData?.meta?.usage ||
    responseData?.usageMetadata ||
    responseData?._meta?.usageMetadata ||
    {};

  // 1. Prompt / Input tokens
  const promptTokens =
    usage.prompt_tokens ??
    usage.promptTokens ??
    usage.input_tokens ??
    usage.inputTokens ??
    usage.promptTokenCount ??
    0;

  // 2. Completion / Output tokens
  const completionTokens =
    usage.completion_tokens ??
    usage.completionTokens ??
    usage.output_tokens ??
    usage.outputTokens ??
    usage.candidatesTokenCount ??
    0;

  // 3. Cached tokens (OpenAI cached_tokens, Anthropic cache_read, Gemini cachedContent)
  const cachedTokens =
    usage.cached_tokens ??
    usage.cachedTokens ??
    usage.prompt_tokens_details?.cached_tokens ??
    usage.cache_read_input_tokens ??
    usage.cachedContentTokenCount ??
    0;

  // 4. Reasoning / Thoughts tokens (OpenAI o1/o3 reasoning_tokens, Gemini 2.0 Flash thoughtsTokenCount)
  const reasoningTokens =
    usage.reasoning_tokens ??
    usage.reasoningTokens ??
    usage.completion_tokens_details?.reasoning_tokens ??
    usage.thoughtsTokenCount ??
    0;

  // 5. Total tokens
  const totalTokens =
    usage.total_tokens ??
    usage.totalTokens ??
    usage.totalTokenCount ??
    (promptTokens + completionTokens);

  const model = responseData?.model || responseData?._meta?.model || undefined;

  return {
    promptTokens,
    completionTokens,
    totalTokens,
    cachedTokens,
    reasoningTokens,
    model
  };
}
